cyw43/src/runner.rs

```diff
@@ -222,6 +274,261 @@ where
+            debug!("BT_CTRL_REG_ADDR = {:08x}", val);
+            if val & BTSDIO_REG_BT_AWAKE_BITMASK != 0 {
+                success = true;
+                break;
+            }
+            Timer::after(Duration::from_millis(1)).await;
+        }
+        assert!(success == true);
+    }
+
+    pub(crate) async fn bt_set_host_ready(&mut self) {
+        debug!("bt_set_host_ready");
+        let old_val = self.bus.bp_read32(HOST_CTRL_REG_ADDR).await;
+        // TODO: do we need to swap endianness on this read?
+        let new_val = old_val | BTSDIO_REG_SW_RDY_BITMASK;
+        self.bus.bp_write32(HOST_CTRL_REG_ADDR, new_val).await;
+    }
+
+    // TODO: use this
+    #[allow(dead_code)]
+    pub(crate) async fn bt_set_awake(&mut self, awake: bool) {
+        debug!("bt_set_awake");
+        let old_val = self.bus.bp_read32(HOST_CTRL_REG_ADDR).await;
+        // TODO: do we need to swap endianness on this read?
+        let new_val = if awake {
+            old_val | BTSDIO_REG_WAKE_BT_BITMASK
+        } else {
+            old_val & !BTSDIO_REG_WAKE_BT_BITMASK
+        };
+        self.bus.bp_write32(HOST_CTRL_REG_ADDR, new_val).await;
+    }
+
+    pub(crate) async fn bt_toggle_intr(&mut self) {
+        debug!("bt_toggle_intr");
+        let old_val = self.bus.bp_read32(HOST_CTRL_REG_ADDR).await;
+        // TODO: do we need to swap endianness on this read?
+        let new_val = old_val ^ BTSDIO_REG_DATA_VALID_BITMASK;
+        self.bus.bp_write32(HOST_CTRL_REG_ADDR, new_val).await;
+    }
+
+    // TODO: use this
+    #[allow(dead_code)]
+    pub(crate) async fn bt_set_intr(&mut self) {
+        debug!("bt_set_intr");
+        let old_val = self.bus.bp_read32(HOST_CTRL_REG_ADDR).await;
+        let new_val = old_val | BTSDIO_REG_DATA_VALID_BITMASK;
+        self.bus.bp_write32(HOST_CTRL_REG_ADDR, new_val).await;
+    }
+
+    pub(crate) async fn init_bt_buffers(&mut self) {
+        debug!("init_bt_buffers");
+        let wlan_ram_base_addr = self.bus.bp_read32(WLAN_RAM_BASE_REG_ADDR).await;
+        assert!(wlan_ram_base_addr != 0);
+        debug!("wlan_ram_base_addr = {:08x}", wlan_ram_base_addr);
+        self.h2b_buf_addr = wlan_ram_base_addr + BTSDIO_OFFSET_HOST_WRITE_BUF;
+        self.b2h_buf_addr = wlan_ram_base_addr + BTSDIO_OFFSET_HOST_READ_BUF;
+        let h2b_buf_in_addr = wlan_ram_base_addr + BTSDIO_OFFSET_HOST2BT_IN;
+        let h2b_buf_out_addr = wlan_ram_base_addr + BTSDIO_OFFSET_HOST2BT_OUT;
+        let b2h_buf_in_addr = wlan_ram_base_addr + BTSDIO_OFFSET_BT2HOST_IN;
+        let b2h_buf_out_addr = wlan_ram_base_addr + BTSDIO_OFFSET_BT2HOST_OUT;
+        self.bus.bp_write32(h2b_buf_in_addr, 0).await;
+        self.bus.bp_write32(h2b_buf_out_addr, 0).await;
+        self.bus.bp_write32(b2h_buf_in_addr, 0).await;
+        self.bus.bp_write32(b2h_buf_out_addr, 0).await;
+    }
+
+    async fn bt_bus_request(&mut self) {
+        // TODO: CYW43_THREAD_ENTER mutex?
```

Comment:
We don't need any mutex here because we cannot get interrupted by the device while doing these operations. (yay cooperative multitasking).

About the interrupt disable, I don't know. If leaving them enabled doesn't corrupt the internal state of the chip, it probably doesn't need to be disabled. We only consume interrupts when we are back on the `select3` call.